Mathlib/LinearAlgebra/FreeModule/ModN.lean

```diff
@@ -0,0 +1,100 @@
+/-
+Copyright (c) 2024 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Algebra.EuclideanDomain.Int
+import Mathlib.Algebra.Module.ZMod
+import Mathlib.LinearAlgebra.Dimension.Free
+
+/-!
+# Quotienting out a free `ℤ`-module
+
+If `G` is a rank `d` free `ℤ`-module, then `G/nG` is a finite group of cardinality `n ^ d`.
+-/
+
+open Finsupp Function
+
+variable {G H M : Type*} [AddCommGroup G] [Module.Free ℤ G] {n : ℕ}
+
+variable (G n) in
+/-- `ModN G n` denotes the quotient of `G` by multiples of `n` -/
+abbrev ModN : Type _ := G ⧸ LinearMap.range (LinearMap.lsmul ℤ G n)
+
+namespace ModN
+
+instance : Module (ZMod n) (ModN G n) := QuotientAddGroup.zmodModule (by simp)
+
+/-- The universal property of `ModN G n` in terms of monoids: Monoid homomorphisms from `ModN G n`
+are the same as monoid homormorphisms from `G` whose values are `n`-torsion. -/
+protected def liftEquiv [AddMonoid M] : (ModN G n →+ M) ≃ {φ : G →+ M // ∀ g, n • φ g = 0} where
+  toFun f := ⟨f.comp (QuotientAddGroup.mk' _), fun g ↦ by
+    let Gn : AddSubgroup G := (LinearMap.range (LinearMap.lsmul ℤ G n)).toAddSubgroup
+    suffices n • g ∈ (QuotientAddGroup.mk' Gn).ker by
+      simp only [AddMonoidHom.coe_comp, comp_apply, ← map_nsmul]
+      change f (QuotientAddGroup.mk' Gn (n • g)) = 0 -- Can we avoid `change`?
+      rw [this, map_zero]
+    simp [QuotientAddGroup.ker_mk', Gn]⟩
+  invFun φ := QuotientAddGroup.lift _ φ <| by rintro - ⟨g, rfl⟩; simpa using φ.property g
+  left_inv f := by
+    ext x
+    induction x using QuotientAddGroup.induction_on
+    rfl -- Should `simp` suffice here?
+  right_inv φ := by aesop
+
+/-- The universal property of `ModN G n` in terms of `ZMod n`-modules: `ZMod n`-linear maps from
+`ModN G n` are the same as monoid homormorphisms from `G` whose values are `n`-torsion. -/
+protected def liftEquiv' [AddCommGroup H] [Module (ZMod n) H] :
+    (ModN G n →ₗ[ZMod n] H) ≃ {φ : G →+ H // ∀ g, n • φ g = 0} :=
+  (AddMonoidHom.toZModLinearMapEquiv n).symm.toEquiv.trans (ModN.liftEquiv )
+
+variable [NeZero n]
+
+/-- Given a free module `G` over `ℤ`, construct the corresponding basis
+of `G / ⟨n⟩` over `ℤ / nℤ`. -/
+noncomputable def basis : Basis (Module.Free.ChooseBasisIndex ℤ G) (ZMod n) (ModN G n) := by
```

Comment:
I think we should generalise slightly to:
```suggestion
noncomputable def basis {ι : Type*} (b : Basis ι ℤ G) : Basis ι (ZMod n) (ModN G n) := by
```
(I tested locally and it is trivial to amend proof for this statement).

It would be nice to add:
```lean
lemma basis_apply_eq_mkQ_apply {ι : Type*} (b : Basis ι ℤ G) (i : ι) :
    basis b i = mkQ n (b i) := by
  sorry
```
but I do not insist.

Reply:
Makes sense, yes! I've also added `basis_apply_eq_mkQ_apply`